benchmarks/linalg_refactor_benchmark.cpp

```diff
@@ -0,0 +1,151 @@
+#include <shogun/base/init.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/linalgrefactor/linalgRefactor.h>
+
+#include <shogun/mathematics/eigen3.h>
+#include <viennacl/linalg/inner_prod.hpp>
+#include <viennacl/vector.hpp>
+
+#include <algorithm>
+#include <memory>
+#include <hayai/hayai.hpp>
+
+#include <iostream>
+
+using namespace shogun;
+
+/**
+ * Instructions :
+ * 1. Install benchmarking toolkit "hayai" (https://github.com/nickbruun/hayai)
+ * 2. Compile against libhayai_main, e.g.
+ * 		g++ -O3 -std=c++11 linglg_refactor_benchmark.cpp -I/usr/include/eigen3 \
+ *		-lshogun -lhayai_main -lOpenCL -o benchmark
+ * 3. ./benchmark
+ */
+
+/** Generate data only once */
+typedef viennacl::vector_base<int32_t, std::size_t, std::ptrdiff_t> VCLVectorBase;
+
+struct Data
+{
+	typedef viennacl::backend::mem_handle VCLMemoryArray;
+
+	Data()
+	{
+		num_rows = 100;
+		A = init(1);
+		B = init(2);
+		Ac = init_c(A);
+		Bc = init_c(B);
+        Ag = init_g(A);
+        Bg = init_g(B);
+		Av = init_v(A);
+		Bv = init_v(B);
+		init_shogun_with_defaults();
+	}
+
+	Data(index_t num_rows)
+	{
+		this->num_rows = num_rows;
+		A = init(1);
+		B = init(2);
+		Ac = init_c(A);
+		Bc = init_c(B);
+		Ag = init_g(A);
+		Bg = init_g(B);
+		Av = init_v(A);
+		Bv = init_v(B);
+		init_shogun_with_defaults();
+	}
+
+	~Data()
+	{
+		exit_shogun();
+	}
+
+	SGVector<int32_t> init(int32_t begin)
+	{
+		SGVector<int32_t> m(num_rows);
+		m.range_fill(begin);
+
+		return m;
+	}
+
+    CPUVector<int32_t> init_c(SGVector<int32_t> m)
+    {
+        CPUVector<int32_t> mc(m);
+        return mc;
+    }
+
+    GPU_Vector<int32_t> init_g(SGVector<int32_t> m)
+    {
+        GPU_Vector<int32_t> mg(m);
+        return mg;
+    }
+
+	VCLVectorBase init_v(SGVector<int32_t> m)
+	{
+		VCLVectorBase mv;
+		std::shared_ptr<VCLMemoryArray> vector(new VCLMemoryArray());
+		viennacl::backend::memory_create(*vector, sizeof(int32_t)*num_rows,
+			viennacl::context());
+		viennacl::backend::memory_write(*vector, 0, num_rows*sizeof(int32_t),
+			m.vector);
+		mv = VCLVectorBase(*vector, num_rows, 0, 1);
+		return mv;
+	}
+
+	SGVector<int32_t> A;
+	SGVector<int32_t> B;
+    CPUVector<int32_t> Ac;
+    CPUVector<int32_t> Bc;
+    GPU_Vector<int32_t> Ag;
+    GPU_Vector<int32_t> Bg;
+	VCLVectorBase Av;
+    VCLVectorBase Bv;
+
+	index_t num_rows;
+};
+
+BENCHMARK_P(CPUVector, dot_int32_explict_eigen3, 10, 1000,
+		(SGVector<int32_t> A, SGVector<int32_t> B))
+{
+	typedef Eigen::Matrix<int32_t, Eigen::Dynamic, 1> VectorXt;
+	Eigen::Map<VectorXt> vec_A(A.vector, A.vlen);
+	Eigen::Map<VectorXt> vec_B(B.vector, B.vlen);
+    //Eigen::Map<VectorXt> vec_B = B;
+    auto C = vec_A.dot(vec_B);
+}
+
+BENCHMARK_P(CPUVector, dot_int32_eigen3, 10, 1000,
+		(CPUVector<int32_t> A, CPUVector<int32_t> B))
+{
+	CPUBackend eigenBackend;
+	sg_linalg->set_cpu_backend(&eigenBackend);
```

Comment:
Backend should be registered in the `Data::init()` itself. Also, I think there should **always** already be a CPU-backend (Eigen). We shouldn't have to set it every time.